### ISSUE TITLE
ValidateAction task leaks metaspace in Gradle Daemon when plugin project has any TaskAction class

### ISSUE THREAD

**Issue by illarionov** [gradle/gradle#24256]
### Steps to Reproduce 
* Сreate a plugin project with some `TransformAction` class (or clone the prepared sample: https://github.com/illarionov/playground-android-composite-build/tree/gradle-transformaction-leak)
* Launch a monitoring tool to track the metaspace used by Gradle Daemon and the number of loaded classes (eg visualvm).
* Run `validatePlugins` task 50—100 times and watch the metaspace usage: 
```
for i in `seq 1000`; do  ./gradlew :plugin:validatePlugins --no-build-cache --no-configuration-cache --rerun; done
```

Number of loaded classes and metaspace size is continuously increasing with each new execution of the task. Our TaskAction class, as well as some other classes, are loaded on every run and are never garbage collected. On a project with a large number of classes this can lead to exhaustion of the metaspace.

If you create and analyze a dump, you will see that the isolated worker-loader `VisitableURLClassLoader` used by the `ValidateAction` task leaks into the field that is reachable from the static instance of the `PropertyValidationAccess` class and therefore it cannot be garbage collected:
![worker-loader](https://raw.githubusercontent.com/illarionov/playground-android-composite-build/gradle-transformaction-leak/img/visitableurlclassloader.png)

After 50 executions, field `PropertyValidationAccess.INSTANCE.typeSchemes[ArtifactTransformActionScheme].inspectionScheme.metadataStore.cache` has 50 copies of the same class loaded by different class loaders and, as it seems, they cannot be unloaded or reused:

![propertyvalidationaccess.png](https://raw.githubusercontent.com/illarionov/playground-android-composite-build/gradle-transformaction-leak/img/propertyvalidationaccess.png)

### Expected behavior
Repeated execution of the `validatePlugins` task should not increase the usage of the metaspace.

### Your Environment
Gradle 8.0.2

**Comment by ov7a** [comment]
Thank you for providing a valid reproducer.

The issue is in the backlog of the relevant team and is prioritized by them.

**Comment by lptr** [comment]
Hm... 🤔 Thanks for the thorough report. I checked out the example project and have been running:

```bash
for i in `seq 1000`; do  ./gradlew :plugin:validatePlugins --no-build-cache --no-configuration-cache --rerun; done
```

I see some moderate increase of metaspace indeed over time:

<img width="1445" alt="image" src="https://user-images.githubusercontent.com/495366/226393993-310682ed-bafe-40eb-b011-9a145f175d07.png">

And I can see the number of `org.gradle.internal.classloader.VisitableURLClassLoader` instances rise in the daemon process, too.

This is weird, because `ValidateAction` is not a task, but a worker API action, and it should execute in a separate daemon process, outside of the main daemon process:

https://github.com/gradle/gradle/blob/cac79ebe509e179e8a45ce0662711a5ae5f16168/subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/ValidatePlugins.java#L82-L100

It is not clear to me how the classes end up in the main daemon process, or why we get this proliferation of `VisitableURLClassLoader`s.

**Comment by illarionov** [comment]
Oh, I noticed that the class loader isolation of this task has been changed since version 8.02 in https://github.com/gradle/gradle/pull/23199.
I rechecked this bug with the new version and I can say that it is not reproducible on Gradle 8.1-rc-1.
Increasing in the number of loaded classes is not as significant as on version 8.02 and the metaspace size decreasing with GC cycles.
I think this issue can be closed, sorry for not checking this on newer version initially.
